src/System.Net.Requests/src/System/Net/WebRequest.cs

```diff
@@ -0,0 +1,492 @@
+
+                            prefixList.Add(new WebRequestPrefixElement("http:", new HttpRequestCreator()));
+                            prefixList.Add(new WebRequestPrefixElement("https:", new HttpRequestCreator()));
+                            prefixList.Add(new WebRequestPrefixElement("file:", new HttpRequestCreator()));
+                            prefixList.Add(new WebRequestPrefixElement("ftp:", new HttpRequestCreator()));
+
+                            s_prefixList = prefixList;
+                        }
+                    }
+                }
+
+                return s_prefixList;
+            }
+            set
+            {
+                s_prefixList = value;
+            }
+        }
+
+        protected WebRequest()
+        {
+        }
+
+        public abstract string Method
+        {
+            get;
+            set;
+        }
+
+        public abstract Uri RequestUri
+        {
+            get;
+        }
+
+        public abstract WebHeaderCollection Headers
+        {
+            get;
+            set;
+        }
+
+        public abstract string ContentType
+        {
+            get;
+            set;
+        }
+
+        public virtual ICredentials Credentials
+        {
+            get
+            {
+                throw NotImplemented.ByDesignWithMessage(SR.net_PropertyNotImplementedException);
+            }
+            set
+            {
+                throw NotImplemented.ByDesignWithMessage(SR.net_PropertyNotImplementedException);
+            }
+        }
+
+        public virtual bool UseDefaultCredentials
+        {
+            get
+            {
+                throw NotImplemented.ByDesignWithMessage(SR.net_PropertyNotImplementedException);
+            }
+            set
+            {
+                throw NotImplemented.ByDesignWithMessage(SR.net_PropertyNotImplementedException);
+            }
+        }
+
+        public abstract IAsyncResult BeginGetResponse(AsyncCallback callback, object state);
+
+        public abstract WebResponse EndGetResponse(IAsyncResult asyncResult);
+
+        public abstract IAsyncResult BeginGetRequestStream(AsyncCallback callback, Object state);
+
+        public abstract Stream EndGetRequestStream(IAsyncResult asyncResult);
+
+        // Offload to a different thread to avoid blocking the caller durring request submission.
+        public virtual Task<Stream> GetRequestStreamAsync()
+        {
+            return Task.Run(() => Task<Stream>.Factory.FromAsync(this.BeginGetRequestStream,
+                this.EndGetRequestStream, null));
+        }
+
+        // Offload to a different thread to avoid blocking the caller durring request submission.
+        public virtual Task<WebResponse> GetResponseAsync()
+        {
+            return Task.Run(() => Task<WebResponse>.Factory.FromAsync(this.BeginGetResponse,
+                this.EndGetResponse, null));
+        }
```

Comment:
There are some perf improvements that can be made to these *Async methods (e.g. avoiding some delegate/closure allocations), but that can be done later.